engine/hatchery/vsphere/spawn.go

```diff
@@ -0,0 +1,224 @@
+}
+
+// createVMModel create a model for a specific worker model
+func (h *HatcheryVSphere) createVMModel(model *sdk.Model) (*object.VirtualMachine, error) {
+	log.Info("Create vm model %s", model.Name)
+	ctx := context.Background()
+	imgCfg := imageConfiguration{}
+
+	if err := json.Unmarshal([]byte(model.Image), &imgCfg); err != nil {
+		return nil, sdk.WrapError(err, "createVMModel> Cannot unmarshal image")
+	}
+
+	vm, errV := h.finder.VirtualMachine(ctx, imgCfg.OS)
+	if errV != nil {
+		return vm, sdk.WrapError(errV, "createVMModel> Cannot find virtual machine")
+	}
+
+	annot := annotation{
+		HatcheryName:            h.Hatchery().Name,
+		WorkerModelLastModified: model.UserLastModified,
+		WorkerModelName:         model.Name,
+		Model:                   true,
+		Created:                 time.Now(),
+	}
+
+	cloneSpec, folder, errCfg := h.createVMConfig(vm, annot)
+	if errCfg != nil {
+		return vm, sdk.WrapError(errCfg, "createVMModel> cannot create VM configuration")
+	}
+
+	task, errC := vm.Clone(ctx, folder, model.Name+"-tmp", *cloneSpec)
+	if errC != nil {
+		return vm, sdk.WrapError(errC, "createVMModel> cannot clone VM")
+	}
+
+	info, errWr := task.WaitForResult(ctx, nil)
+	if errWr != nil || info.State == types.TaskInfoStateError {
+		return vm, sdk.WrapError(errWr, "createVMModel> state in error")
+	}
+
+	vm = object.NewVirtualMachine(h.vclient.Client, info.Result.(types.ManagedObjectReference))
+
+	if _, errW := vm.WaitForIP(ctx); errW != nil {
+		return vm, sdk.WrapError(errW, "createVMModel> cannot get an ip")
+	}
+
+	if _, errS := h.launchClientOp(vm, imgCfg.UserData+"; shutdown -h now", nil); errS != nil {
+		log.Warning("createVMModel> cannot start program %s", errS)
+		annot := annotation{ToDelete: true}
+		if annotStr, err := json.Marshal(annot); err == nil {
+			vm.Reconfigure(ctx, types.VirtualMachineConfigSpec{
+				Annotation: string(annotStr),
+			})
+		}
+	}
+
+	ctxTo, cancel := context.WithTimeout(ctx, 4*time.Minute)
+	defer cancel()
+	if err := vm.WaitForPowerState(ctxTo, types.VirtualMachinePowerStatePoweredOff); err != nil {
+		return nil, sdk.WrapError(err, "createVMModel > cannot wait for power state result")
+	}
+	log.Info("createVMModel> model %s is build", model.Name)
+
+	modelFound, errM := h.getModelByName(model.Name)
+	if errM == nil {
+		if errD := h.deleteServer(modelFound); errD != nil {
+			log.Warning("createVMModel> Cannot delete previous model %s : %s", model.Name, errD)
+		}
+	}
+
+	task, errR := vm.Rename(ctx, model.Name)
+	if errR != nil {
+		return vm, sdk.WrapError(errR, "createVMModel> Cannot rename model %s", model.Name)
+	}
+
+	if _, err := task.WaitForResult(ctx, nil); err != nil {
+		return vm, sdk.WrapError(err, "createVMModel> error on waiting result for vm renaming %s", model.Name)
+	}
+
+	// FIXME doesn't work, give me a forbidden
+	return vm, h.client.WorkerModelUpdate(model.ID, model.Name, model.Type, model.Image, cdsclient.WorkerModelOpts.WithoutRegistrationNeed())
```

Comment:
this func create a "vm snapshot" for a worker model, but why does it need to update worker model? A worker model does not need to know if a "snapshot vm" exists or not.

Reply:
Certainly a misunderstanding about flag `needRegistration`. In fact in my mind `needRegistration` is the flag that tell me the worker model was updated or need to create a new snapshot  in order to be update. So in this case my problem was to set `needRegistration` to false after creating a new snapshot/model. Am I wrong with this assessment ?

Reply:
needRegistration = ask worker model to run again "worker register", to let worker send all capabilities to API. In same func, Hatchery snaphosts the VM to make an image, but it's a specific use case of openstack & vsphere, because there is no flag "NeedRegistration Because User Update The Worker Model"
So, Hatchery OPenstack & VSphere does not need to update worker model when there create a image.

Reply:
Ok right thanks ! So if I need to know if I have to create a new snapshot/model I must check the date and not this flag. :)

Reply:
yes. The openstack code of that:


```go
// if it's was a worker model for registration
// check if we need to create a new openstack image from it
// by comparing userDateLastModified from worker model
```

https://github.com/ovh/cds/blob/master/engine/hatchery/openstack/openstack.go#L196